conmon/conmon.c

```diff
@@ -504,52 +504,96 @@ static void conn_sock_shutdown(int how)
+static gboolean stdio_cb(int fd, GIOCondition condition, gpointer user_data);
+
+static gboolean tty_hup_timeout_scheduled = false;
+
+static gboolean tty_hup_timeout_cb (G_GNUC_UNUSED gpointer user_data)
+{
+	tty_hup_timeout_scheduled = false;
+	g_unix_fd_add (masterfd_stdout, G_IO_IN, stdio_cb, GINT_TO_POINTER(STDOUT_PIPE));
+	return G_SOURCE_REMOVE;
+}
+
+static bool read_stdio(int fd, stdpipe_t pipe, bool *eof)
 {
 	#define STDIO_BUF_SIZE 8192 /* Sync with redirectResponseToOutputStreams() */
 	/* We use one extra byte at the start, which we don't read into, instead
 	   we use that for marking the pipe when we write to the attached socket */
 	char real_buf[STDIO_BUF_SIZE + 1];
         char *buf = real_buf + 1;
-	stdpipe_t pipe = GPOINTER_TO_INT(user_data);
 	ssize_t num_read = 0;
 
-	if ((condition & G_IO_IN) != 0) {
-		num_read = read(fd, buf, BUF_SIZE);
-		if (num_read < 0) {
-			nwarn("stdio_input read failed %s", strerror(errno));
+	if (eof)
+		*eof = false;
+
+	num_read = read(fd, buf, STDIO_BUF_SIZE);
+	if (num_read == 0) {
+		if (eof)
+			*eof = true;
+		return false;
+	} else if (num_read < 0) {
+		nwarn("stdio_input read failed %s", strerror(errno));
+		return false;
+	} else {
+		if (write_k8s_log(log_fd, pipe, buf, num_read) < 0) {
+			nwarn("write_k8s_log failed");
 			return G_SOURCE_CONTINUE;
 		}
 
-		if (num_read > 0) {
-			if (write_k8s_log(log_fd, pipe, buf, num_read) < 0) {
-				nwarn("write_k8s_log failed");
-				return G_SOURCE_CONTINUE;
-			}
+		real_buf[0] = pipe;
+		if (conn_sock_writable && write_all(conn_sock, real_buf, num_read+1) < 0) {
+			nwarn("Failed to write to socket");
+			conn_sock_shutdown(SHUT_WR);
+		}
+		return true;
+	}
+}
 
-                        real_buf[0] = pipe;
-			if (conn_sock_writable && write_all(conn_sock, real_buf, num_read+1) < 0) {
-				nwarn("Failed to write to socket");
-				conn_sock_shutdown(SHUT_WR);
-			}
+static gboolean stdio_cb(int fd, GIOCondition condition, gpointer user_data)
+{
+	stdpipe_t pipe = GPOINTER_TO_INT(user_data);
+	bool did_read = false;
+	bool read_eof = false;
+
+	/* When we get here, condition can be G_IO_IN and/or G_IO_HUP.
+	   IN means there is some data to read.
+	   HUP means the other side closed the fd. In the case of a pine
+	   this in final, and we will never get more data. However, in the
+	   terminal case this just means that nobody has the terminal
+	   open at this point, and this can be change whenever someone
+	   opens the tty */
+
+	/* Read any data before handling hup */
+	if ((condition & G_IO_IN) != 0) {
+		did_read = true;
+		read_stdio(fd, pipe, &read_eof);
+	}
 
-			return G_SOURCE_CONTINUE;
+	if ((condition & G_IO_HUP) && opt_terminal && (read_eof || !did_read)) {
+		/* We got a HUP from the terminal, this means there
+		   are no open slaves ptys atm, and we will get a lot
+		   of wakeups until we have one, switch to polling
+		   mode. */
+		if (!tty_hup_timeout_scheduled) {
+			g_timeout_add (100, tty_hup_timeout_cb, NULL);
```

Comment:
it will wait 100msec, then add the fd again. The it will get HUP error again, remove the fd and do a new timeout. Until there is a new slave, then it will be as normal.